File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def preorderTraversal(self, root):
        ans = []
        if root is None:
            return ans
        
        while root:
            if root.left:
                curr = root.left
                while curr.right and curr.right != root:
                    curr = curr.right
                
                if curr.right:
                    curr.right = None
                    root = root.right
                else:
                    ans.append(root.val)
                    curr.right = root
                    root = root.left
            else:
                ans.append(root.val)
                root = root.right

        return ans
        

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ans;
        TreeNode* curr;

        if(!root) return ans;

        while(root) {
            if(root -> left) {
                curr = root -> left;
                while(curr -> right && curr -> right != root) {
                    curr = curr -> right;
                }

                if(curr -> right) {
                    curr -> right = NULL;
                    root = root -> right;
                }
                else {
                    ans.push_back(root -> val);
                    curr -> right = root;
                    root = root -> left;
                }
            }
            else {
                ans.push_back(root -> val);
                root = root -> right;
            }
        }

        return ans;
    }
};